ddtrace/internal/wrapping/context.py

```diff
@@ -0,0 +1,686 @@
+            push_exc_info
+            push_null
+            load_const                  {context_exit}
+            precall                     0
+            call                        0
+            pop_top
+            reraise                     2
+        tried
+
+        _except:
+            copy                        3
+            pop_except
+            reraise                     1
+        """
+    )
+
+elif sys.version_info >= (3, 10):
+    CONTEXT_HEAD.parse(
+        r"""
+            load_const                  {context}
+            setup_with                  @_except
+            pop_top
+        _except:
+        """
+    )
+
+    CONTEXT_RETURN.parse(
+        r"""
+            load_const                  {context}
+            load_method                 $__return__
+            rot_three
+            rot_three
+            call_method                 1
+        """
+    )
+
+    CONTEXT_FOOT.parse(
+        r"""
+            with_except_start
+            pop_top
+            reraise                     1
+        """
+    )
+
+elif sys.version_info >= (3, 9):
+    CONTEXT_HEAD.parse(
+        r"""
+            load_const                  {context}
+            setup_with                  @_except
+            pop_top
+        _except:
+        """
+    )
+
+    CONTEXT_RETURN.parse(
+        r"""
+            load_const                  {context}
+            load_method                 $__return__
+            rot_three
+            rot_three
+            call_method                 1
+        """
+    )
+
+    CONTEXT_FOOT.parse(
+        r"""
+            with_except_start
+            pop_top
+            reraise
+        """
+    )
+
+
+elif sys.version_info >= (3, 7):
+    CONTEXT_HEAD.parse(
+        r"""
+            load_const                  {context}
+            setup_with                  @_except
+            pop_top
+        _except:
+        """
+    )
+
+    CONTEXT_RETURN.parse(
+        r"""
+            load_const                  {context}
+            load_method                 $__return__
+            rot_three
+            rot_three
+            call_method                 1
+        """
+    )
+
+    CONTEXT_FOOT.parse(
+        r"""
+            with_cleanup_start
+            with_cleanup_finish
+            end_finally
+            load_const                  None
+            return_value
+        """
+    )
+
+
+# This is abstract and should not be used directly
+class BaseWrappingContext(t.ContextManager):
+    __priority__: int = 0
+    __frame__: t.Optional[FrameType] = None
+
+    def __init__(self, f: FunctionType):
+        self.__wrapped__ = f
+
+    @classmethod
+    def wrapped(cls, f: FunctionType) -> "BaseWrappingContext":
+        if cls.is_wrapped(f):
+            context = cls.extract(f)
+            assert isinstance(context, cls)  # nosec
+        else:
+            context = cls(f)
+            context.wrap()
+        return context
+
+    def __return__(self, value):
```

Comment:
Should we try to combine `__exit__` and `__return__`?

I am thinking about cases like finishing spans, we now will always have two code paths needed to close a span or context for these wrappers.

e.g. doesn't seem like we have a `try/finally` type setup here which we could get with the other wrapping we do.


`def __exit__(self, return_value, (exc_type, exc_val, exc_tb)): -> return_value`


It does seem to conflate things a bit, but wondering if we can consolidate a bit.


```python
class MyWrapper(WrappingContext):
    def __enter__(self, frame):
        self.span = tracer.trace("name")

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.span.set_exc_info(exc_type, exc_val, exc_tb)
        self.span.finish()

    def __return__(self, rv):
        self.span.finish()
        return rv
```




```python
class MyWrapper(WrappingContext):
    def __enter__(self, _frame):
        self.span = tracer.trace("name")

    # __finally__ get return value and exception info
    # __finally__'s return value is ignored
    def __finally__(self, rv, (exc_type, exc_val, exc_tb)):
        self.span.set_exc_info(exc_type, exc_val, exc_tb)
        self.span.finish()


    # Only define `__return__` if you want to intercept and modify the return value
```


Not sure I love this, just starting conversation to get your thoughts.

Reply:
I was thinking that one could do something like
```
def _close_context(self, return_value=None, exc_info=None):
    ...

def __return__(self, value):
    self._close_context(return_value=value)

def __exit__(self, *exc_info):
    self._close_context(exc_info=exc_info)
```
in a subclass of `WrappingContext` if there is a lot that can be DRYed in general

Reply:
yeah, we can always subclass, ok... makes sense.